Cleanup PDL::IO::FITS tests, fix incorrect tests; fix regression in writing 0D ndarrays

djerius opened the pull request:
`t/fits.t` reorganization and fixes:

- many tests shared variables, requiring explicit re-initialization
- some regression tests for legacy FITS header writing were incorrectly excluded if `Astro::FITS::HEADER` was installed
- the test for tilde expansion in file names was never run.
- temporary files were explicitly unlinked instead of using the automatic facilities of `File::Temp`

Tests are now isolated into subtests, with no shared variables. The other issues were dealt with, but the level of header checking with and without `Astro::FITS::Header` is still quite asymmetric.

bug fixes:

- null hdus were always written using the legacy Perl code due to an incorrect check if `Astro::FITS::Header` was available.

regression fixes:

- passing 0D ndarrays to `wfits` caused an internal crash, as the code
  expected they be at least 1D. This used to work.

djerius committed:
use subtests to modularize/isolate tests.  use File::Temp auto-unlinking; fix temp file race conditions; fix tilde expansion test;

- Tests are now better isolated by using subtests and not reusing
  variables between tests

- ile::Temp's automatic file unlinking is now used, rather than
  explicitly unlinking the temp files

bugfixes;

- use of unique temporary file names could fail due to race conditions.

  File::Temp creates the temporary file and returns a handle to it.
  The file is detroyed when the handle is destroyed, and the filename
  can then be reused by another process.

  The code kept the name of the temporary file, but not the handle, so
  the file was destroyed before the code could use it, and thus
  a race condition could have ensued.

- the tilde expansion tests did not actually run

  The tests should be skipped if the '~' expanded directory was not
  writeable.  The test for that was essentially

      -w '~'

  which checks for writeability of a path which is exactly the
  character '~', not the expanded path.

djerius committed:
add test for regression: exception when writing scalar ndarrays

djerius committed:
test that FITS null hdu is written using correct legacy/non-legacy code path

djerius committed:
use correct name of flag used to determine if Astro::FITS::Header was available

djerius committed:
regression fix: write fits files from scalar ndarrays

mohawk2 commented on t/fits.t:
```diff
@@ -2,340 +2,480 @@ use strict;
+};
+
+###############################
+subtest 'write null hdu with and without Astro::FITS::Header' => sub {
```
The existence of `t/fits-noafh.t` renders this whole section moot. In particular, merely checking that `wfits` didn't throw an exception seems to miss the point a bit. Please remove this change, or make it unconditional on AFH being there, and do an actual check of the read-back data.

djerius replied:
It does do an actual check of the readback data.  the test for exception is secondary to the check which appears on the next lines.

mohawk2 replied:
OK. Then please get rid of this whole bit, since exceptions make tests fail anyway.

mohawk2 commented on t/fits.t:
```diff
@@ -2,340 +2,480 @@ use strict;
+            is( $$table2{hdr}{TTYPE2}, "COLB", "column #2 is COLB" );
+            is( $$table2{hdr}{TFORM2}, "1D",   "  stored as 1D" );
+
+            is_pdl $x, $$table2{COLA}, "COLA";
```
These two have the "got" and "expected" values the wrong way round. Please fix this.

djerius replied:
ok.

mohawk2 commented on t/fits.t:
```diff
@@ -2,340 +2,480 @@ use strict;
+                rfits $file;
+            };
+
+            ok(
```
Please do here the multiple checks you did above. A single "not OK" would be a rather useless test diagnostic.

mohawk2 commented on t/fits.t:
```diff
@@ -2,340 +2,480 @@ use strict;
+            ##my $f = float(1,0,-1,2) + i * float( 0,1,2,-1 );
+
+            my $table2 = do {
+                my $table = {
```
This intermediate variable that's used once is not adding value. Please just call `wfits` with that value.

djerius replied:
ok

mohawk2 commented on t/fits.t:
```diff
@@ -2,340 +2,480 @@ use strict;
+                rfits $file;
+            };
+
+            ok(
```
Multiple checks again please.

djerius replied:
yup.

mohawk2 commented on t/fits.t:
```diff
@@ -2,340 +2,480 @@ use strict;
+                $p->wfits($file);
+                $q = PDL->rfits($file);
+                my $flag = 1;
+                if ( ${ $p->get_dataref } ne ${ $q->get_dataref } ) {
```
A simple `is` here would suffice. Please use that.

djerius replied:
ok

mohawk2 commented on t/fits.t:
```diff
@@ -2,340 +2,480 @@ use strict;
+  SKIP: {
+        my $tilde_dir = <~>;
+
+        skip '~ directory is not writeable'
```
This is an incorrect call to `skip`.

djerius replied:
Looks like Test2's skip doesn't require the count, which is why it passed on my end. fixed.

mohawk2 reviewed (changes requested):
Please make the changes noted. It's a bit surprising that you submitted this without (from the CI results) even running it locally.

djerius commented:
> Please make the changes noted. It's a bit surprising that you submitted this without (from the CI results) even running it locally.

Harsh, dude. No need for that.

djerius committed:
legacy Perl header reader did not handle multiple comments correctly

djerius committed:
order of got,expected was reversed

djerius committed:
split complex test into several

djerius committed:
rewrite unnecessarily complex test

djerius committed:
rewrite in light of t/fits-noah.t, read header instead of searching the raw fits file

Unify with/without Astro::FITS::Header tests for writing nullhdu, as
patht/fits-noah.t takes care of running without Astro::FITS::Header.

Scanning the raw fits file for expected COMMENT strings results in horrendous
garbage output if the tests fail. Instead use rfitshdr to extract
the header and scan the comments.

djerius committed:
writing bintables is not supported without Astro::FITS::Header

djerius committed:
let t/fits-noaht.t take care of the case where Astro::FITS::Header is not present

djerius committed:
non Test2 provided skip requires the number of tests skipped

djerius committed:
simplify

djerius commented:
> > regression fixes:
> > 
> > * passing 0D ndarrays to `wfits` caused an internal crash, as the code expected they be at least 1D. This used to work.
> 
> Can you confirm in which version that used to work? It will need noting in `Changes` (which I'll update).

I know with certainty it worked back in 2.018. I'll try to track down the version where it changed.

djerius commented:
> > > regression fixes:
> > > 
> > > * passing 0D ndarrays to `wfits` caused an internal crash, as the code expected they be at least 1D. This used to work.
> > 
> > 
> > Can you confirm in which version that used to work? It will need noting in `Changes` (which I'll update).
> 
> I know with certainty it worked back in 2.018. I'll try to track down the version where it changed.

git bisect says 899a6159a60478dd0d197a1623981237efade1d6.

git run script:

```
git reset --hard
perl Makefile.PL
make -j80 && perl -Mblib -MPDL -MPDL::IO::FITS=wfits -E 'wfits { a => pdl(3)}, q{/tmp/foo}'
```
